src/variables.jl

```diff
@@ -600,10 +581,8 @@ function set_integer(variable_ref::VariableRef)
         error("Cannot set the variable_ref $(variable_ref) to integer as it " *
               "is already binary.")
     end
-    constraint_ref = MOI.add_constraint(backend(owner_model(variable_ref)),
-                                        MOI.SingleVariable(index(variable_ref)),
-                                        MOI.Integer())
-    set_integer_index(variable_ref, constraint_ref)
+    MOI.add_constraint(backend(owner_model(variable_ref)),
```

Comment:
The `is_integer` and `is_binary` above might be the performance bottleneck. It should be replaced by errors at the MOI level like:
https://github.com/JuliaOpt/MathOptInterface.jl/blob/3f4857816281a79744af0170391c435e318123f1/src/Utilities/model.jl#L431-L445

Reply:
Why is `is_valid` expensive? I thought it would just look up an element in an array.

Reply:
`is_integer` called `backend` which is type unstable. Maybe we should have `moi_set_integer` etc... to make sure we call `backend` only once in `JuMP.add_variable`